text/0004-configurational-dependencies.md

```diff
@@ -0,0 +1,55 @@
+# Configurational dependencies
+
+## Summary
+
+A new type of dependencies for installing configuration and hooks.
+
+## Motivation
+
+We want to make it possible to share some configurations between projects. Storing the configuration in regular dependencies is not an option as we might need the configuration during installation of the "regular" dependencies ("dependencies", "devDependencies", and "optionalDependencies"). Hence, we need to install these configurational dependencies before other types of dependencies.
+
+Some examples of usage:
+
+* Installing hooks used by [`.pnpmfile.cjs`](https://pnpm.io/pnpmfile).
+* Installing the list of dependencies that are allowed to be built (the [pnpm.onlyBuiltDependenciesFile](https://pnpm.io/package_json#pnpmonlybuiltdependenciesfile)).
+* Installing [catalogs](https://pnpm.io/catalogs).
+* Loading patch file.
+
+## Detailed Explanation
+
+There will be a new field in `package.json` called `pnpm.configDependencies`. For example:
+
+```json
+{
+  "name": "my-pkg",
+  "version": "0.0.0",
+  "pnpm": {
+    "configDependencies": {
+      "my-configs": "1.0.0+sha512-30iZtAPgz+LTIYoeivqYo853f02jBYSd5uGnGpkFV0M3xOt9aN73erkgYAmZU43x4VfqcnLxW9Kpg3R5LC4YYw=="
+    }
+  }
+}
+```
+
+These new type of "configurational" dependencies will be npm packages with a lot of limitations:
+
+* They won't have any dependencies. Even if they will have dependencies, pnpm will ignore them during installation.
+* They will not have lifecycle scripts.
+* They will only be installable via exact versions.
+
+## Rationale and Alternatives
+
+For the list of "onlyBuiltDependencies" currently we can load them from `node_modules`. However, that does introduce some "works on my machine" issues as we don't have list during the linking from store stage. As a result, if the dependency is not allowed to be built, we still might load it from the side-effects cache.
+
+## Implementation
+
+These dependencies will be installed as early as possible in order to be able to load settings from it. So this should happen ouside of the `@pnpm/core` module. When we call `mutateModules` we should already have all the configurations from `pnpm.configDependencies` loaded and ready.
+
+## Prior Art
+
```

Comment:
Is this similar to [`build-dependencies`](https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#build-dependencies) in Rust?

Reply:
I believe this is different than `build-dependencies` in the Rust ecosystem.

This RFC is similar what [templates](https://github.com/pnpm/rfcs/pull/3) was trying to do with reusable configuration, but only for pnpm specific settings. Similar to templates, configuration is stored in the package manifest since that's available at earlier stages of the installation process. However, I think it's a lot simpler than templates and likely to be less controversial since it's only for pnpm related settings and not existing `package.json` fields.